### ISSUE TITLE
Allow to specify a read action for a relationship in a load

### ISSUE THREAD

**Issue by vonagam** [ash-project/ash#711]
This is a feature request for a convenient way to specify a read action for a load of a relationship.

Example: Let's say we have a blog with Post and User. Post has `belongs_to :author, User`. And there are two relevant pages - a page with a list of all posts and a page for a single post. In both cases when posts are read we want to load their authors too. But in a list an author is represented only by a name while on a post page we want to show their avatar, small bio/about paragraph, maybe posts count, stuff like that.

The problem is that currently there is no simple way to do it.

A possible solution is to allow passing action name in load with syntax equal to that for calculations:
```elixir
query |> load(relationship: :action)
query |> load(relationship: [action: args])
query |> load(relationship: [action: {args, nested_loads}])
```

Thoughts?

**Comment by zachdaniel** [comment]
You can currently do `load(relationship: Ash.Query.for_read(Resource, :action))`, IIRC. Would need to test how that behavior plays with the current logic, as at the moment relationships are explicitly configured with an action.

i.e

```elixir
relationships do
  has_many :foos, Foo do
    read_action :foo
  end
end
```

**Comment by zachdaniel** [comment]
You can also configure two actions w/ different names that use a different read action

**Comment by zachdaniel** [comment]
I'm mostly concerned about the syntax of it, as the options you showed are not unambiguous (if you had a relationship or calculation called the same name as the action)

**Comment by zachdaniel** [comment]
but you can't have a relationship or calculation or attribute with the same name. I think if we want to support it we'd need to come up with some kind of non-ambuguous name. We have a reserved name `:as`, that can be used to load things with a different name i.e `relationship: [as: :foo, ...]`. Perhaps we could repurpose this, i.e with `as: {:name, :action}`, but even that is a bit messy. I'm open to the idea, just need unambiguous syntax.

**Comment by vonagam** [comment]
> but you can't have a relationship or calculation or attribute with the same name.

Well, Ash 3 is coming and and I don't see why action names cannot be introduced into that uniqueness check. I assume words like `read`, `create`, `update`, `delete`, `get` and so on are not really used for those other things, so there won't be many collisions.

**Comment by zachdaniel** [comment]
Hm....that is true. But is this particular feature worth the introduction of that requirement? For instance, those other things can't share a name because they live in the same map, it's a technical impossibility for them to share a name. I think I'd want to see something like what you laid out, or perhaps even a new pattern around loading things that is cleaner even, like `[relationship_name: Ash.Load.action(:action)]` or something (thats not the right answer, just making something up).

**Comment by vonagam** [comment]
Yeah, same, I don't know the perfect solution at the moment.

Another thing which I thought about adding possibly later and for more generic purposes is to accept functions to customize a relationship query:
```elixir
query |> load(relationship: customize)
query |> load(relationship: {customize, nested_loads})
```
Where `customize` accepts and returns a query. (For example `for_read` inside action definition might not work because it will not have access to `actor` and other query opts, but with `customize` those opts can be present and read from the input query.)

**Comment by vonagam** [comment]
Another option (again, just throwing things at the board) is to use tuples:
```elixir
query |> load(relationship: {:action})
query |> load(relationship: {:action, args})
query |> load(relationship: {:action, args, nested_loads})
```
This should be unambiguous. But I'm not sure about the look (nothing against, just not sure).

**Comment by zachdaniel** [comment]
Yeah, a map could work. Currently if you see a map in a load statement it can only be calculation arguments I believe. I'm open to it 👍

**Comment by nallwhy** [comment]
With Ash 4.0 in mind, what about unifying the load syntax for both relationships and calculations using a map-based approach?

Currently the syntax is inconsistent:

```elixir
# relationship
load(author: [:posts])
load(author: custom_query)

# calculation - map means args, tuple for args + nested load
load(full_name: %{separator: "-"})
load(full_name: {%{separator: "-"}, [:nested]})
```

Proposal: keyword value as a list always means a nested load (for both relationships and calculations), and a map provides detailed options:

```elixir
# simple — list is always nested load
load(:author)
load(:full_name)
load(author: [:posts])
load(full_name: [:nested_field])

# detailed — map with explicit keys, same shape for both
load(author: %{action: :detailed, load: [:posts]})
load(full_name: %{args: %{separator: "-"}, load: [:nested]})
```

This would:
- Naturally solve this issue — `action:` becomes just another map key
- Remove the ad-hoc `{args, nested_load}` tuple syntax for calculations
- Make the map semantics unambiguous (currently `calc: %{...}` looks like it could be a load statement but it's actually args)
- Be extensible for future options (e.g., `sort:`, `filter:`)

The main trade-off is a breaking change for calculation argument passing (`calc: %{arg: val}` → `calc: %{args: %{arg: val}}`), but a major version bump would be the right time for it.

**Comment by zachdaniel** [comment]
Hmm...it's an interesting proposal for sure. I like that it's very regular. I think even for 4.0 it would be really difficult to find all the places that do this wrong and update them, but I can't think of any other reasonable way to do it that isn't breaking 🤔
